[wang-h-z] iP

wang-h-z opened the pull request:
### Friday
> "TGIF." - Katy Perry

Can't seem to remember your schedule? Friday's got your back. Friday is: 
* easy to use
* ~easy~ SUPER EASY to learn
* **FAST** :fire:

How to use Friday: 
1. clone the repo from [here](https://github.com/wang-h-z/ip)
2. follow the README.md file 
3. enjoy!

Best part about Friday? Its completely **FREE!**

Features: 
- [x] Managing tasks
- [ ] Filtering tasks (coming soon)
- [ ] Finding tasks (coming soon) 

Here's a short snippet of Friday's codebase! 

```java
public static void main(String[] args) {
        Friday bot = new Friday("data/ui.Friday.txt");
        bot.start();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

wang-h-z committed:
Level 0. Rename, Greet, Exit

There are several methods to display the greeting exit messages. I have decided to store the greetings message and exit message as separate strings "header" and "ending". This improves the formatting of the code and makes future edits easier.

wang-h-z committed:
Add UserInputHandler class

Create a UserInputHandler class that will handle the logic for reading user input and generating appropriate responses.

wang-h-z committed:
Add UserInputHandler logic

UserInputHandler uses Scanner object from java.util to communicate with user. Scanner object is initialised when UserInputHandler() constructor is called.

UserInputHandler logic has two basic methods getInput() and handleInput() to receive input from the user and to handle the input from the user.

Friday creates UserInputHandler object to handle user input when Friday is being used.

wang-h-z committed:
Add BotMessage class, added BotMessage to Friday logic and added basic chatting logic

Used BotMessage class to encapsulate a bot response from Friday. Created a method start() in Friday to start a chat with a user. The chat ends only when the user inputs the String "bye". This is the intended behaviour of Level 1.

wang-h-z committed:
Add basic list functionality to chatbot

Add ArrayList field to Friday which will store the inputs of the user. Add printlist() method to Friday which handles the logic for generating the string responsible for displaying the stored user inputs when user prompts the bot with "list". Adjusted start() logic to handle the prompt "list". Added additional String formatting to handleInput() in UserInputHandler.java.

wang-h-z committed:
Create Task class

Task class encapsulates the user input by storing the task description and an additional boolean field to determine the status of the Task. Add markAsDone() and markAsUndone() methods to change the isDone state. Change ArrayList in Friday to store Task objects instead. Add mark and unmark user input logic to start() in Friday.

wang-h-z committed:
Add Todo class

Add Todo class. Testing SourceTree push changes.

wang-h-z committed:
Add Deadline and Event

Add Deadline and Event classes. Testing Git settings again.

wang-h-z committed:
Add inheritance

Deadline, Event and Todo classes inherit from Task class.

wang-h-z committed:
Modify start() logic to work with new Task subclasses and improved bot response

Use indexOf() and substring() methods to extract text from user input assuming that the user gives Friday an appropriate response. Use trim() to remove any additional whitespaces after the input description. This streamlines editing for the bot response.

wang-h-z committed:
Modify runtest.sh to work with Friday

Changed file destinations to use runtest.sh script. Additionally, added spacings and "\t" tab spaces to beautify Friday's messages.

wang-h-z committed:
Exception handling for Todo and Incorrect input

Add exception handling for Todo and incorrect inputs. Will add exception tests to scripts.sh and also add more exception handling to reach typical and stretch goals level.

wang-h-z committed:
Delete command

Added delete command which deletes an entry according to the index of the Task in the list. For future reference, add exception handling for when trying to remove an item which is out of bounds of the array length (1, list.size()).

wang-h-z committed:
Exception handling for deadline class

Tweak the usage of TodoException to a more generalised DescriptionException class as all 3 current Task subclasses contain a description and can be given an invalid description. Added a DeadlineException class to handle invalid deadlines after /by command. Finally, added logic to handle invalid description and invalid deadline given to a deadline command. Additionally, segmented the code in handleInput() for future reference using simple comments.

wang-h-z committed:
Exception handling for Event input

Add 3 exceptions to handle 3 different invalid inputs. MissingCommandException for missing /from and /to commands. FromException and ToException ensures that there is no empty inputs for those fields. Tweaked the logic of reading a deadline task to conform with MissingCommandException.

wang-h-z committed:
Exception handling for delete command

Added basic exception handling for delete command which resolves invalid inputs to a delete command, trying to delete an item from an empty list and also trying to delete an item which does not exist in the list.

wang-h-z committed:
Add Enums to UserInputHandler

Changed if else logic to use enums instead. Using enums allows the usage of switch case which improves readability and maintainability of the code base compared to if else statements. Moved printList() method to UserInputHandler as it made sense that UserInputHandler should provide this feature instead of Friday. Finally, moved more input cases into UserInputHandler.

wang-h-z committed:
Add save function

Friday saves Tasks into a .txt file. Previous entries into the list is also recorded and loaded when Friday runs again. Storage class handles the saving of Tasks in the .txt file. Storage contains two methods, loadTasks() and saveTasks(). loadTasks() allows previously logged tasks to be recorded and used. saveTasks() allows new entries to be added and removed from the list, while being logged into the .txt file. Added logic to all subclasses of Task and Task to facilitate loadTasks().

wang-h-z committed:
Merge branch-Level-7 into master

wang-h-z committed:
Date and time in Friday

Friday now understands and is able to interpret some formatted string inputs utilising java.time.LocalDateTime and java.time.format.DateTimeFormatter. Developers can change accepted formats in the FORMATTERS array in DateTimeParser.java. Friday also accepts a new command known as DATE. A DATE command takes in one of the accepted formatted string inputs and returns a list of deadline or event tasks that need to be done by the string input date.

wang-h-z committed:
Abstract Friday functionality

HandleUserInput renamed to Parser for better understandability. Ui class is now responsible for reading user inputs and also responding to user inputs via System.out in the terminal. TaskList abstraction to encapsulate the original ArrayList<Task> to store the list of tasks. Storage class now receives a filepath String in its constructor, increasing usability.

wang-h-z committed:
Command class and command subclasses

Added further abstraction by encapsulating user input processing to an abstract command class. Commands can be executed using execute(), which handles the logic to add a task to the TaskList, handles exceptions if any and also returns a ui output via the terminal. This simplifies the Parser class, increasing readability.

wang-h-z committed:
Organise classes into packages

Segmented classes and subclasses into various packages following addressbook-level2 as guideline. Added relevant import statements into file headers so that code functions with new file structure.

wang-h-z committed:
Merge branch 'add-gradle-support'

wang-h-z committed:
JUnit testing

Achieved basic JUnit testing for Friday. Tested successfully TodoTest. Used stubs to test successfully ParserTest.

wang-h-z committed:
Add JavaDoc to all non-trivial methods

JavaDoc has been added to all non-trivial methods in all classes.

wang-h-z committed:
Fix import statements

Fixed import statements to conform to Java coding standard.

wang-h-z committed:
Add header to Friday

wang-h-z committed:
Add Find command

Added FindCommand class to execute a keyword search on the tasks in the project.

wang-h-z committed:
Merge tag 'A-CodingStandard'

wang-h-z committed:
Merge tag 'A-JavaDoc'

Dinoman44 commented on src/main/java/common/BotMessage.java:
```diff
@@ -0,0 +1,15 @@
+
+    @Override
+    public String toString() {
+        return '\t' + this.spacing + '\n' + this.description + '\n' + '\t' + this.spacing + "\n";
```
Small nit, you could have just done `... this.description + '\n\t' + ...` but eh.

Dinoman44 commented on src/main/java/exceptions/FridayException.java:
```diff
@@ -0,0 +1,8 @@
+package exceptions;
```
Seems like all these exceptions are related to Task handling. Why don't you create a [sub-package](https://www.refreshjava.com/java/sub-packages-in-java) inside of the `tasks` package so you can then import the exceptions as `import tasks.exceptions.<exception>`?

Dinoman44 commented on src/main/java/tasks/Todo.java:
```diff
@@ -0,0 +1,29 @@
+     * @param str The String that represents the Todo Task.
+     * @return
+     */
+    public static Todo fromString(String str) {
```
Small nit, you seem to be missing the `@Override` annotation here.

Dinoman44 commented on src/main/java/tasks/Event.java:
```diff
@@ -0,0 +1,57 @@
+            this.fromDate = DateTimeParser.parse(from);
+        } catch (IllegalArgumentException e) {
+            //No correct format, deadline only recorded as a String
+            this.fromDate = null;
```
Would it be a better idea to reprompt the user for input instead of setting the `fromDate`/`toDate` as `null`? This way working with DateTimes would be easier and you can allow for more functionality (like searching for tasks on a day).

Dinoman44 commented on src/main/java/tasks/Event.java:
```diff
@@ -0,0 +1,57 @@
+     * @return
+     */
+    public static Event fromString(String str) {
+        boolean isDone = str.contains("[X]");
```
What if some weird case inputs `"[X]"` as the task description? It might be better to confirm that the `"[X]"` is at the right place (i.e. from index 3 to 5) if at all present.

Dinoman44 commented on src/main/java/tasks/Task.java:
```diff
@@ -0,0 +1,51 @@
+        }
+    }
+
+    public String getDescription() {
```
Small nit: again missing `@Override` annotation.

Dinoman44 commented on src/main/java/common/DateTimeParser.java:
```diff
@@ -0,0 +1,53 @@
+package common;
```
Since this is a parser, why not have it under the `parser` package instead?

Dinoman44 commented on src/main/java/common/TaskList.java:
```diff
@@ -0,0 +1,44 @@
+package common;
```
Since this is related to tasks, why not have it in the `tasks` package?

Dinoman44 commented on src/main/java/ui/Ui.java:
```diff
@@ -0,0 +1,91 @@
+        return this.scanner.nextLine();
+    }
+
+    public void listOutput(TaskList list) {
```
Naming note: It's better to have function names as verbs, ex: `outputList` or `outputTaskAddition`.

Dinoman44 commented on src/main/java/commands/ByeCommand.java:
```diff
@@ -0,0 +1,30 @@
+     */
+    @Override
+    public void execute(TaskList list, Ui ui, Storage storage) {
+        ui.showEnd();
```
I may have missed it, but where is the part where the tasks get stored from the tasklist to the text file?

Dinoman44 commented on src/main/java/commands/DateCommand.java:
```diff
@@ -0,0 +1,76 @@
+        ui.dateOutput(userTime.format(formatter), list);
+    }
+
+    /**
```
For overridden functions I think you can just have the JavaDoc comment in the top-level class, then no need to re-define it for the classes that inherit from that class.

Dinoman44 commented on src/main/java/commands/ByeCommand.java:
```diff
@@ -0,0 +1,30 @@
+    /**
+     * The ByeCommand signifies that the user would like to stop the chatbot.
+     *
+     * @return false
```
```suggestion
     * @return true
```

Dinoman44 commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,71 @@
+    public void saveTasks(TaskList list) {
+        try (PrintWriter writer = new PrintWriter(new FileWriter(FILE_PATH))) {
+            for (Task task : list.getList()) {
+                writer.println(task.toString());
```
I did not know you could do that, I might try this out in my code. I was using the `java.io.FileWriter` class and the `FileWriter::write` method.

Dinoman44 commented on data/Friday.txt:
```diff
@@ -0,0 +1,7 @@
+[T][ ] Task
+[D][ ] Task (by: 23-02-2020 1800)
+[D][ ] Task (by: Feb 06 2002, 6:00 pm)
+[T][ ] wth
```
```suggestion
[T][X] wth
```

Dinoman44 commented on build.gradle:
```diff
@@ -0,0 +1,41 @@
+}
+
+shadowJar {
+    archiveBaseName = "duke"
```
```suggestion
    archiveBaseName = "friday"
```

Dinoman44 commented on build.gradle:
```diff
@@ -0,0 +1,41 @@
+    archiveClassifier = null
+}
+
+run{
```
```suggestion
run {
```

Dinoman44 reviewed (commented):
Looks solid, some minor refactoring and coding convention changes needed. Good luck!

wang-h-z committed:
Add CheckStyle to Friday

Slowly fix CheckStyle errors in the future.

ulricolo7 commented on src/main/java/tasks/Deadline.java:
```diff
@@ -0,0 +1,49 @@
+
+public class Deadline extends Task {
+
+    private String by;
+    private LocalDateTime byDate;
```
Would it be better if the variables here are named more descriptively, instead of 'by' and 'byDate'?

ulricolo7 commented on src/main/java/tasks/Event.java:
```diff
@@ -0,0 +1,57 @@
+import common.DateTimeParser;
+
+public class Event extends Task {
+    private String from;
+    private String to;
+    private LocalDateTime fromDate;
+    private LocalDateTime toDate;
```
Similar comment for 'from', 'to', 'fromDate', and 'toDate'! Would it be better if it was named more descriptively?

ulricolo7 commented on src/main/java/ui/Friday.java:
```diff
@@ -0,0 +1,62 @@
+import storage.Storage;
+
+public class Friday {
+    private final BotMessage header = new BotMessage("""
```
Perhaps add a descriptor as to what header this is for? The variable 'ending' too

ulricolo7 reviewed (commented):
LGTM! Just pointing out some naming conventions! Generally could be a bit more descriptive haha

wang-h-z committed:
Add GUI to Friday

Followed javafx-tutorial to add basic working GUI to Friday.

wang-h-z committed:
Resources for Friday GUI

wang-h-z committed:
Update GUI error message

Made a small adjustment to how error messages are displayed in Friday GUI. Changed the indentation for GUI to be consistent to other GUI message bubbles from Friday. Made a small adjustment to the ui/Friday class such that the text UI for Friday still works as intended.

wang-h-z committed:
Add assertions to Friday

Basic assertions to ensure that exceptions caused by user mis-inputs are handled appropriately in the logic of creating Task classes.

wang-h-z committed:
Merge pull request #1 from wang-h-z/A-Assertions

Add assertions to Friday

wang-h-z committed:
Ensure code quality standards are met

Enforce SLAP and prominent code throughout codebase. Moved some variables for better readability.

wang-h-z committed:
Add Stream logic to TaskList

Changed findTasks to implement java Streams.

wang-h-z committed:
Simplify stream logic

wang-h-z committed:
Merge pull request #2 from wang-h-z/A-Streams

A-Streams

wang-h-z committed:
Merge pull request #3 from wang-h-z/A-CodeQuality

A-CodeQuality

wang-h-z committed:
Setup CI workflow to Friday

wang-h-z committed:
Improve runtest and Friday GUI logic

Encountered issues during CI integration which led to the discovery that there were testing issues with runtest.sh script. After noticing issues with the runtest.sh script, noticed another mistake in the the way Friday GUI was loading tasks. It has been fixed and should be working as intended. Additionally, added the missing welcome message to Friday GUI.

wang-h-z committed:
Adjust filepath for runtest.bat

wang-h-z committed:
Increase timeout and modify runtest

wang-h-z committed:
Filepath change to runtest

wang-h-z committed:
Improve filepath for runtest

wang-h-z committed:
Modify CI workflow and runtest

wang-h-z committed:
Modify runtest.bat

wang-h-z committed:
Modify runtest

wang-h-z committed:
Update runtest

wang-h-z committed:
Runtest update

wang-h-z committed:
Runtest update

wang-h-z committed:
Update runtest

wang-h-z committed:
Update runtest

wang-h-z committed:
Update runtest

wang-h-z committed:
Update runtest

wang-h-z committed:
Update CI/CD workflow

wang-h-z committed:
Merge pull request #11 from wang-h-z/A-CI

A-CI increment

wang-h-z committed:
Priority command

Added priority command to Friday's available commands. Every task now contains a field which indicates the priority of a task. H stands for high, M stands for medium and L stands for low.

The priority command requires an index (according to the index shown in the TaskList) and a relevant priority field after the /lvl command to assign a priority to a task.

wang-h-z committed:
Update Expected.txt for runtest.sh

wang-h-z committed:
Merge pull request #12 from wang-h-z/C-Priority

C-Priority increment

wang-h-z committed:
GUI improvements to Friday

Made overall improvements to the Friday GUI.

Changed the Friday and User icons to circular icons to
make them more appealing. Added a title to the GUI as
requested by the module to differentiate from Duke. Also
added promptText to the userInput bar to help guide users
to interact with it. Improved the dialog box by introducing
text bubbles with margins and borders. Additionally, styled
the send button and added a background to the chat.

wang-h-z committed:
Merge pull request #14 from wang-h-z/A-BetterGUI

A Better GUI

wang-h-z committed:
Change Unkown Command message

wang-h-z committed:
Update runtest.bat

wang-h-z committed:
Merge pull request #15 from wang-h-z/A-BetterGUI

Change Unkown Command message

wang-h-z committed:
Improve welcome message

wang-h-z committed:
Merge pull request #16 from wang-h-z/A-Personality

Improve welcome message

wang-h-z committed:
Improve JUnit tests

Added new EventTest to test Event task class. Added some get() methods in Task to facilitate testing.

wang-h-z committed:
Fixed JUnit bugs

wang-h-z committed:
DateTimeParser format consistency

wang-h-z committed:
Merge pull request #17 from wang-h-z/A-MoreTesting

Improve JUnit tests

wang-h-z committed:
Update README

Additionally, made some bug fixes at the same time.

wang-h-z committed:
Merge pull request #18 from wang-h-z/A-UserGuide

Update README

wang-h-z committed:
Update EventTest

wang-h-z committed:
Refactor Friday Commands

Most of the Friday Commands experience repeat code due to having to generate both
a GUI response and CLI response. The logic is mostly consistent but each response
require a different return value.

Refactored most of the Friday Commands to reduce this repetitiveness. Additionally,
utilised SLAP to improve code quality.

wang-h-z committed:
Refactor exceptions

Removed some of the subclass exceptions that can just be replaced by FridayException.
This simplifies the code for exception-handling, allowing for better readability
and understandability. Code quality is improved.

wang-h-z committed:
Abstract String processing

Applied SLAP to the Task classes to hide the input String processing behind a layer
of abstraction to improve code quality.

wang-h-z committed:
Improve JavaDocs

Made improvements to most if not all src files in Friday. Attempted to normalise parameters definition
across similar files in the same package.

wang-h-z committed:
Event task error handling

Added error handling for an edge case where the /from DateTimeObject is
later than the /to DateTimeObject, which is an error that should be raised
to the user.

wang-h-z committed:
Merge tag 'A-CheckStyle'

wang-h-z committed:
Update gradle.yml

Clear gradle cache to deal with TimeZone discrepancies which could be causing CI workflow to fail.

wang-h-z committed:
Remove unusual JUnit test

JUnit test for new error handling works in local gradle build but does not work
in CI workflow. Temporarily removed until a better fix for it can be found.

wang-h-z committed:
Merge pull request #19 from wang-h-z/A-MoreErrorHandling

More Error Handling

wang-h-z committed:
Merge branch 'master' of github.com:wang-h-z/ip

wang-h-z committed:
Fix Checkstyle errors

Rectified all checkstyle errors in src/main classes. Rectified most checkstyle
errors in src/test classes.

wang-h-z committed:
Rectify Checkstyle errors

Corrected the last src/test Checkstyle errors.

wang-h-z committed:
Add JUnit tests

Added EventCommandTest, DeleteCommandTest and Deadline Test for
further JUnit testing.

wang-h-z committed:
Update JUnit test

Unusual CI workflow behaviour occured. Local gradle tests are passing
but not CI tests. Removed these irregular tests for now.

wang-h-z committed:
Merge pull request #20 from wang-h-z/A-MoreTesting

Add JUnit tests

wang-h-z committed:
Update README

Updated README and also cleared Friday.txt to facilitate
release version.

wang-h-z committed:
Merge pull request #21 from wang-h-z/A-Release

Update README

wang-h-z committed:
Update README

wang-h-z committed:
Update README

Fixed some minor formatting and spelling errors.
